clients/cobol-lsp-vscode-extension/src/provider/TarCopybookFileSystemProvider.ts

```diff
@@ -0,0 +1,200 @@
+import { Minimatch } from "minimatch";
+import * as vscode from "vscode";
+import { Memoize } from "../services/util/Memoize";
+import { TarUtil } from "../services/util/TarUtil";
+
+export class TarCopybookFileSystemProvider
+  implements vscode.FileSystemProvider
+{
+  public static readonly SCHEME = "cobol-ls-tar";
+  public tarCache = new Memoize(
+    async (tarFileUri: vscode.Uri) => {
+      return await TarUtil.readTarFile(tarFileUri);
+    },
+    undefined,
+    (tarFileUri: vscode.Uri) => {
+      const { tarfsPath, dialect } = this.getDetailsFromTarUri(tarFileUri);
+      return `${tarfsPath}$$${dialect}`;
+    },
+  );
+  private emitter = new vscode.EventEmitter<vscode.FileChangeEvent[]>();
+  readonly onDidChangeFile: vscode.Event<vscode.FileChangeEvent[]> =
+    this.emitter.event;
+
+  private watchers = new Set<WatchEntry>();
+
+  private fireChange(event: vscode.FileChangeEvent) {
+    this.emitter.fire([event]);
+
+    if (event.type === vscode.FileChangeType.Deleted) {
+      this.clearCache();
+    }
+  }
+  watch(
+    resource: vscode.Uri,
+    opts: { recursive: boolean; excludes: string[] },
+  ): vscode.Disposable {
+    const entry: WatchEntry = { resource, opts, disposed: false };
+    this.watchers.add(entry);
+
+    return new vscode.Disposable(() => {
+      entry.disposed = true;
+      this.watchers.delete(entry);
+    });
+  }
+
+  /**
+   * `${TarFileContentProvider.SCHEME}://${tarFileUri.fsPath}?filePath=${evaluatedInternalPath}#${dialect}`
+   * @param uri
+   * @returns
+   */
+  async stat(uri: vscode.Uri): Promise<vscode.FileStat> {
+    const { tarfsPath, filePath } = this.getDetailsFromTarUri(uri);
+    const matchingFiles = await this.fetchMatchingFiles(
+      tarfsPath,
+      uri,
+      filePath,
+    );
+    if (matchingFiles && matchingFiles[0]) {
+      return matchingFiles[0].fileData.fileMetadata;
+    }
+    throw vscode.FileSystemError.FileNotFound();
+  }
+
+  /**
+   * `${TarFileContentProvider.SCHEME}://${tarFileUri.fsPath}?searchPath=${evaluatedInternalPath}#${dialect}`
+   * get list of all the files within a TAR file and search pattern
+   * @param uri
+   */
+  async readDirectory(uri: vscode.Uri): Promise<[string, vscode.FileType][]> {
+    const { tarfsPath, searchPath } = this.getDetailsFromTarUri(uri);
+    const tarContent = await this.tarCache.execute(vscode.Uri.file(tarfsPath));
+    if (!tarContent) {
+      throw new Error(`Failed to retrieve content for URI: ${uri.toString()}`);
+    }
+    const result: [string, vscode.FileType][] = [];
+    const matchingFiles = tarContent.filter((e) =>
+      new Minimatch(searchPath + "/*", {
+        nocase: true,
+        dot: true,
+      }).match(e.fileName),
+    );
+    matchingFiles.forEach((t) =>
+      result.push([t.fileName, t.fileData.fileMetadata.type]),
+    );
+    return result;
+  }
+
+  createDirectory(uri: vscode.Uri): void | Thenable<void> {
+    this.fireChange({ type: vscode.FileChangeType.Created, uri });
```

Comment:
This needs to work the other way around... The implementation should throw, because it does not allow external modifications, but we should internally fire the event after we `TarUtil.readTarFile`.